server/app/service/package.scala

```diff
@@ -10,14 +12,28 @@ package object service {
     def check = output.filter(!_.contains(testFailedMarker))
   }
 
-  def testResult(report: Try[String]): TestResult =
-    report match {
-      case Success(s) =>
-        val status =
-          if (s.contains(testFailedMarker)) TestStatus.FailedByTest
-          else TestStatus.Passed
-        TestResult(status.toString)
-      case Failure(e) =>
-        TestResult(TestStatus.FailedByCompilation.toString, e.getMessage)
+  type CheckNext = (String) => Unit
+
+  type OnBlockComplete = (Try[Unit]) => Unit
+
+  type PushResult = (Event) => Unit
+
+  def test(pushResult: PushResult): (CheckNext, OnBlockComplete) = {
+    def isFailed(line: String): Boolean = line.startsWith(red) && line.contains("*** FAILED ***")
+
+    def testResult(blockRes: Try[Unit], failedByTest: => Boolean): TestResult = blockRes match {
+      case Success(_) if failedByTest => TestResult(TestStatus.FailedByTest.toString)
+      case Success(_) => TestResult(TestStatus.Passed.toString)
+      case Failure(e) => TestResult(TestStatus.FailedByCompilation.toString, e.getMessage)
     }
+
+    val cancelable = BooleanCancelable()
+
+    val checkNext = (line: String) => if (isFailed(line)) cancelable.cancel()
```

Comment:
`checkNext` is shortest but still meaningful name, lets finalize it on later stages